File: polyglot/python/profile_merger.py
#!/usr/bin/env python3
"""
polyglot/python/profile_merger.py

Merges multiple GitHub profile configuration files into a single, valid output.
Supports JSON, YAML, and TOML formats with deep merge semantics.
"""

import argparse
import json
import os
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional

try:
    import yaml
except ImportError:
    yaml = None

try:
    import toml
except ImportError:
    toml = None


@dataclass
class MergeConfig:
    """Configuration for the merge operation."""
    
    input_files: List[str] = field(default_factory=list)
    output_file: str = "merged_profile.json"
    format: str = "json"  # json, yaml, or toml
    overwrite: bool = False
    deep_merge: bool = True
    conflict_strategy: str = "last_wins"  # last_wins, first_wins, custom
    
    def __post_init__(self):
        if not self.input_files:
            raise ValueError("At least one input file must be specified")


def load_yaml_file(filepath: Path) -> Dict[str, Any]:
    """Load a YAML file and return its contents as a dictionary."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = yaml.safe_load(f)
        if content is None:
            return {}
        return content


def load_toml_file(filepath: Path) -> Dict[str, Any]:
    """Load a TOML file and return its contents as a dictionary."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = toml.load(f)
        if not isinstance(content, dict):
            return {}
        return content


def load_json_file(filepath: Path) -> Dict[str, Any]:
    """Load a JSON file and return its contents as a dictionary."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = json.load(f)
        if not isinstance(content, dict):
            return {}
        return content


def detect_format(filepath: Path) -> str:
    """Detect the format of a file based on extension."""
    ext = filepath.suffix.lower()
    mapping = {'.yaml': 'yaml', '.yml': 'yaml', 
               '.json': 'json', '.toml': 'toml'}
    return mapping.get(ext, 'json')


def deep_merge(base: Dict[str, Any], override: Dict[str, Any], 
               strategy: str = "last_wins") -> Dict[str, Any]:
    """
    Perform a deep merge of two dictionaries.
    
    Args:
        base: The base dictionary (existing profile)
        override: The overriding dictionary (new values)
        strategy: Conflict resolution strategy
        
    Returns:
        A new merged dictionary
    """
    result = dict(base)
    
    for key, value in override.items():
        if key not in result:
            # New key - just add it
            result[key] = value
        elif isinstance(result[key], dict) and isinstance(value, dict):
            # Both are dicts - recurse deeper
            result[key] = deep_merge(result[key], value, strategy)
        else:
            # Conflict - apply strategy
            if strategy == "last_wins":
                result[key] = value
            elif strategy == "first_wins":
                pass  # Keep original
            elif strategy == "custom":
                raise ValueError("Custom strategy requires a callback function")
    
    return result


def validate_profile(profile: Dict[str, Any]) -> List[str]:
    """
    Validate that the profile contains required fields.
    
    Returns a list of validation errors (empty if valid).
    """
    errors = []
    
    # Required top-level keys for GitHub profiles
    required_keys = ['username', 'name']
    
    missing = [key for key in required_keys if key not in profile]
    if missing:
        errors.append(f"Missing required fields: {', '.join(missing)}")
    
    # Validate types
    if 'email' in profile and not isinstance(profile['email'], str):
        errors.append("'email' must be a string")
    
    if 'bio' in profile and (not isinstance(profile['bio'], str) or 
                            len(profile['bio']) > 2000):
        pass  # Allow bio to be optional
    
    return errors


def load_all_files(config: MergeConfig) -> List[Dict[str, Any]]:
    """Load all input files and return their contents."""
    loaded = []
    
    for filepath in config.input_files:
        path = Path(filepath)
        
        if not path.exists():
            print(f"Warning: File not found - {path}")
            continue
        
        fmt = detect_format(path)
        
        try:
            if fmt == 'yaml':
                data = load_yaml_file(path)
            elif fmt == 'toml':
                data = load_toml_file(path)
            else:
                data = load_json_file(path)
            
            loaded.append(data)
            print(f"Loaded: {path} ({fmt})")
        except Exception as e:
            print(f"Error loading {path}: {e}")
    
    return loaded


def merge_profiles(loaded: List[Dict[str, Any]], 
                   config: MergeConfig) -> Dict[str, Any]:
    """Merge all loaded profiles into one."""
    if not loaded:
        return {}
    
    result = dict(loaded[0])
    
    for data in loaded[1:]:
        result = deep_merge(result, data, 
                          config.conflict_strategy)
    
    return result


def save_profile(profile: Dict[str, Any], output_path: Path, 
                 fmt: str) -> None:
    """Save the merged profile to a file."""
    with open(output_path, 'w', encoding='utf-8') as f:
        if fmt == 'yaml':
            yaml.dump(profile, f, default_flow_style=False, sort_keys=True)
        elif fmt == 'toml':
            import tomli_w
            toml_str = tomli_w.dumps(profile)
            f.write(toml_str)
        else:  # json
            json.dump(profile, f, indent=2, ensure_ascii=False, sort_keys=True)


def main() -> int:
    """Main entry point for the CLI."""
    parser = argparse.ArgumentParser(
        description="Merge multiple GitHub profile configuration files."
    )
    
    parser.add_argument("files", nargs="+", 
                       help="Input config files to merge")
    parser.add_argument("-o", "--output", default="merged_profile.json",
                       help="Output file path (default: merged_profile.json)")
    parser.add_argument("--format", choices=["json", "yaml", "toml"],
                       default="json", help="Output format")
    parser.add_argument("--no-deep-merge", action="store_true",
                       help="Shallow merge instead of deep merge")
    parser.add_argument("-v", "--verbose", action="store_true",
                       help="Enable verbose output")
    
    args = parser.parse_args()
    
    config = MergeConfig(
        input_files=args.files,
        output_file=args.output,
        format=args.format,
        deep_merge=not args.no_deep_merge,
    )
    
    # Load all files
    loaded = load_all_files(config)
    
    if not loaded:
        print("No valid profiles found to merge.")
        return 1
    
    # Merge them
    merged = merge_profiles(loaded, config)
    
    # Validate
    errors = validate_profile(merged)
    if errors:
        print("\nValidation warnings:")
        for error in errors:
            print(f"  - {error}")
    
    # Save output
    output_path = Path(args.output)
    save_profile(merged, output_path, config.format)
    
    print(f"\nMerged profile saved to: {output_path.resolve()}")
    
    if args.verbose and errors:
        return 1
    
    return 0


if __name__ == "__main__":
    sys.exit(main() or 0)

File: polyglot/go/profile_merger.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// Profile represents a GitHub profile structure.
type Profile struct {
	Name         string            `json:"name,omitempty" yaml:"name,omitempty"`
	DisplayName  string            `json:"display_name,omitempty" yaml:"display_name,omitempty"`
	Bio          string            `json:"bio,omitempty" yaml:"bio,omitempty"`
	URL          string            `json:"url,omitempty" yaml:"url,omitempty"`
	Email        string            `json:"email,omitempty" yaml:"email,omitempty"`
	Location     string            `json:"location,omitempty" yaml:"location,omitempty"`
	Socials      map[string]string `json:"socials,omitempty" yaml:"socials,omitempty"`
	Tags         []string          `json:"tags,omitempty" yaml:"tags,omitempty"`
	Metadata     Metadata          `json:"metadata,omitempty" yaml:"metadata,omitempty"`
}

// Metadata holds internal/optional profile data.
type Metadata struct {
	SourceFiles   []string `json:"source_files,omitempty" yaml:"source_files,omitempty"`
	MergedAt      string   `json:"merged_at,omitempty" yaml:"merged_at,omitempty"`
	Version       string   `json:"version,omitempty" yaml:"version,omitempty"`
}

// MergeOptions controls merge behavior.
type MergeOptions struct {
	AppendArrays    bool // If true, arrays are appended instead of replaced
	DryRun          bool // If true, print to stdout only
	OutputFormat    string // "yaml", "json", or auto-detect
	SortTags        bool   // Sort tags alphabetically after merge
}

// DefaultOptions returns sensible defaults.
func DefaultOptions() MergeOptions {
	return MergeOptions{
		AppendArrays:  false,
		DryRun:        false,
		OutputFormat:  "yaml",
		SortTags:      true,
	}
}

// LoadProfile reads a profile from file (YAML or JSON).
func LoadProfile(path string) (*Profile, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading %s: %w", path, err)
	}

	var profile Profile
	contentType := detectContentType(data)

	switch contentType {
	case "yaml":
		err = yaml.Unmarshal(data, &profile)
	case "json":
		err = json.Unmarshal(data, &profile)
	default:
		return nil, fmt.Errorf("unknown format for %s", path)
	}

	if err != nil {
		return nil, fmt.Errorf("parsing %s as %s: %w", path, contentType, err)
	}

	// Preserve source file info if not present
	if profile.Metadata.SourceFiles == nil {
		profile.Metadata.SourceFiles = []string{path}
	} else {
		// Avoid duplicates
		existing := make(map[string]bool)
		for _, f := range profile.Metadata.SourceFiles {
			existing[f] = true
		}
		if !existing[path] {
			profile.Metadata.SourceFiles = append(profile.Metadata.SourceFiles, path)
		}
	}

	return &profile, nil
}

// detectContentType determines if content is YAML or JSON.
func detectContentType(data []byte) string {
	trimmed := strings.TrimSpace(string(data))
	if len(trimmed) == 0 {
		return "yaml" // default fallback
	}

	// Check for JSON indicators
	if strings.HasPrefix(trimmed, "{") || strings.HasPrefix(trimmed, "[") {
		var temp interface{}
		if err := json.Unmarshal(data, &temp); err == nil {
			return "json"
		}
	}

	// Default to YAML (more common for profiles)
	return "yaml"
}

// MergeProfiles combines multiple profiles into one.
func MergeProfiles(profiles []*Profile, opts MergeOptions) (*Profile, error) {
	if len(profiles) == 0 {
		return nil, fmt.Errorf("no profiles provided")
	}

	result := &Profile{}

	// Initialize result with defaults from first profile
	result = profiles[0]

	for _, p := range profiles[1:] {
		mergeInto(result, p, opts)
	}

	// Apply post-merge processing
	if opts.SortTags && len(result.Tags) > 0 {
		sort.Strings(result.Tags)
	}

	return result, nil
}

// mergeInto performs a deep merge of two profiles.
func mergeInto(base, overlay *Profile, opts MergeOptions) {
	// Name fields: last one wins
	if overlay.Name != "" {
		base.Name = overlay.Name
	}
	if overlay.DisplayName != "" {
		base.DisplayName = overlay.DisplayName
	}

	// Bio: append if non-empty and not already set
	if overlay.Bio != "" && base.Bio == "" {
		base.Bio = overlay.Bio
	} else if overlay.Bio != "" && base.Bio != "" {
		// If both have bios, overlay wins (or could concatenate)
		base.Bio = overlay.Bio
	}

	// URL: last one wins
	if overlay.URL != "" {
		base.URL = overlay.URL
	}

	// Email: last one wins
	if overlay.Email != "" {
		base.Email = overlay.Email
	}

	// Location: last one wins
	if overlay.Location != "" {
		base.Location = overlay.Location
	}

	// Socials: deep merge maps
	if base.Socials == nil && overlay.Socials != nil {
		base.Socials = make(map[string]string)
		for k, v := range overlay.Socials {
			base.Socials[k] = v
		}
	} else if base.Socials != nil && overlay.Socials != nil {
		for k, v := range overlay.Socials {
			if _, exists := base.Socials[k]; !exists || opts.AppendArrays {
				base.Socials[k] = v
			}
		}
	}

	// Tags: smart merge based on options
	if len(overlay.Tags) > 0 {
		if opts.AppendArrays && len(base.Tags) == 0 {
			base.Tags = append([]string(nil), overlay.Tags...)
		} else if !opts.AppendArrays || len(base.Tags) == 0 {
			base.Tags = append([]string(nil), overlay.Tags...)
		} else {
			// Append only new tags, avoid duplicates
			existing := make(map[string]bool)
			for _, t := range base.Tags {
				existing[t] = true
			}
			for _, t := range overlay.Tags {
				if !existing[t] {
					base.Tags = append(base.Tags, t)
					existing[t] = true
				}
			}
		}
	}

	// Metadata: merge source files and timestamps
	if base.Metadata.SourceFiles == nil && overlay.Metadata.SourceFiles != nil {
		base.Metadata.SourceFiles = make([]string, 0, len(overlay.Metadata.SourceFiles))
		for _, f := range overlay.Metadata.SourceFiles {
			base.Metadata.SourceFiles = append(base.Metadata.SourceFiles, f)
		}
	} else if base.Metadata.SourceFiles != nil && overlay.Metadata.SourceFiles != nil {
		existing := make(map[string]bool)
		for _, f := range base.Metadata.SourceFiles {
			existing[f] = true
		}
		for _, f := range overlay.Metadata.SourceFiles {
			if !existing[f] {
				base.Metadata.SourceFiles = append(base.Metadata.SourceFiles, f)
				existing[f] = true
			}
		}
	}

	// Set merge timestamp if not present
	if base.Metadata.MergedAt == "" {
		import "time"
		base.Metadata.MergedAt = time.Now().Format(time.RFC3339)
	}
}

// Import: need to fix the import above - let me restructure
func main() {
	var opts MergeOptions = DefaultOptions()
	var inputs []string
	var output string

	// Parse command-line arguments
	for _, arg := range os.Args[1:] {
		if strings.HasPrefix(arg, "--") {
			switch arg {
			case "--help", "-h":
				printUsage()
				os.Exit(0)
			case "--dry-run", "-d":
				opts.DryRun = true
			case "--format=":
				format := strings.TrimPrefix(arg, "--format=")
				if format != "yaml" && format != "json" {
					fmt.Fprintf(os.Stderr, "Invalid format. Use 'yaml' or 'json'.\n")
					os.Exit(1)
				}
				opts.OutputFormat = format
			case "--sort-tags":
				opts.SortTags = true
			case "--no-sort-tags":
				opts.SortTags = false
			default:
				fmt.Fprintf(os.Stderr, "Unknown option: %s\n", arg)
				printUsage()
				os.Exit(1)
			}
		} else if strings.HasPrefix(arg, "-") {
			switch arg {
			case "--help":
				printUsage()
				os.Exit(0)
			case "-d":
				opts.DryRun = true
			default:
				inputs = append(inputs, arg)
			}
		} else if !strings.HasPrefix(arg, "-") && !opts.DryRun {
			if output == "" {
				output = arg
			} else {
				fmt.Fprintf(os.Stderr, "Multiple output files not supported yet.\n")
				os.Exit(1)
			}
		} else if opts.DryRun {
			inputs = append(inputs, arg)
		}
	}

	if len(inputs) == 0 {
		fmt.Fprintf(os.Stderr, "No input files specified.\n\n")
		printUsage()
		os.Exit(1)
	}

	if output == "" && !opts.DryRun {
		output = "merged_profile.yaml"
	}

	// Load all profiles
	var profiles []*Profile
	for _, path := range inputs {
		absPath, err := filepath.Abs(path)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error resolving %s: %v\n", path, err)
			os.Exit(1)
		}

		p, err := LoadProfile(absPath)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Warning: %v\n", err)
			continue // Continue with remaining files
		}

		profiles = append(profiles, p)
		fmt.Printf("Loaded: %s\n", absPath)
	}

	if len(profiles) == 0 {
		fmt.Fprintf(os.Stderr, "No valid profiles found.\n")
		os.Exit(1)
	}

	// Merge profiles
	merged, err := MergeProfiles(profiles, opts)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Merge error: %v\n", err)
		os.Exit(1)
	}

	// Output result
	var outData []byte
	switch opts.OutputFormat {
	case "json":
		outData, _ = json.MarshalIndent(merged, "", "  ")
	default: // yaml
		outData, _ = yaml.Marshal(merged)
	}

	if opts.DryRun {
		fmt.Println(string(outData))
	} else {
		err = os.WriteFile(output, outData, 0644)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error writing output: %v\n", err)
			os.Exit(1)
		}
		fmt.Printf("Merged profile saved to: %s\n", output)

		// Print summary
		printSummary(merged)
	}
}

func printUsage() {
	fmt.Println(`
cognis-digital profile-merger - Merge GitHub profile configurations

Usage:
  cognis-digital merge [options] <file1.yaml> <file2.json> ...

Options:
  -d, --dry-run       Print merged result to stdout without writing
  --format=TYPE       Output format: yaml (default) or json
  --sort-tags         Sort tags alphabetically after merge
  --no-sort-tags      Keep original tag order
  -h, --help          Show this help message

Examples:
  # Merge two YAML files and save output
  cognis-digital merge profile1.yaml profile2.yaml > merged.yaml

  # Dry run to preview result
  cognis-digital merge -d profile1.yaml profile2.yaml

  # Merge with JSON output
  cognis-digital merge --format=json profile1.yaml profile2.yaml
`)
}

func printSummary(p *Profile) {
	fmt.Println("\n=== MERGED PROFILE SUMMARY ===")
	if p.Name != "" {
		fmt.Printf("Name: %s\n", p.Name)
	}
	if p.DisplayName != "" {
		fmt.Printf("Display Name: %s\n", p.DisplayName)
	}
	if p.Bio != "" {
		fmt.Printf("Bio: %s\n", p.Bio)
	}
	if p.URL != "" {
		fmt.Printf("URL: %s\n", p.URL)
	}
	if p.Email != "" {
		fmt.Printf("Email: %s\n", p.Email)
	}
	if p.Location != "" {
		fmt.Printf("Location: %s\n", p.Location)
	}

	if len(p.Socials) > 0 {
		fmt.Println("\nSocial Links:")
		for k, v := range p.Socials {
			fmt.Printf("  - %s: %s\n", k, v)
		}
	}

	if len(p.Tags) > 0 {
		fmt.Printf("\nTags (%d): ", len(p.Tags))
		fmt.Print(strings.Join(p.Tags, ", "))
	}

	if !p.Metadata.MergedAt == "" {
		fmt.Printf("\nMerged at: %s\n", p.Metadata.MergedAt)
	}

	fmt.Println("==============================")
}